README.md
GNT: The GLib Ncurses Toolkit
=============================

GNT is an ncurses toolkit for creating text-mode graphical user interfaces in a
fast and easy way. It is based on [GLib](https://wiki.gnome.org/Projects/GLib)
and [ncurses](https://www.gnu.org/software/ncurses/ncurses.html).

It was born out of the console-based UI, **Finch**, for the [libpurple
project](https://developer.pidgin.im/wiki/WhatIsLibpurple), but has now been
split into its own independent repository.

Building GNT
------------

To build, you will need [Meson](https://mesonbuild.com/),
[GLib](https://wiki.gnome.org/Projects/GLib),
[ncurses](https://www.gnu.org/software/ncurses/ncurses.html),
[libxml2](http://xmlsoft.org/), and [Python 3](https://www.python.org/). Exact
versions can be determined from the `meson.build` in the top-level directory.

On Debian-based systems, install `meson` `ninja-build` `gobject-introspection`
`libgirepository1.0-dev` `gtk-doc-tools` `libglib2.0-dev` `libxml2-dev`
`libncurses-dev` `libpython3-dev`.

On Fedora-based systems, install `meson` `ninja-build` `gobject-introspection`
`gtk-doc` `glib2-devel` `libxml2-devel` `ncurses-devel` `python3-devel`.

You can then run Meson and Ninja as usual (passing any `-D` options if necessary):

```bash
$ meson build
$ ninja -C build
$ ninja -C build install
```

Notes About Versioning
----------------------

This code was originally contained in the Pidgin/Finch/libpurple repository.
Versions were set by the top-level `configure.ac` in that repository. The GNT
version was independent of Pidgin and Finch and the tag. This repository
contains the original tags and so you may see multiple tags pointing to the
same commit. The GNT version *as it was released* is noted in the second column
of the table below.

A second `configure.ac` was included in the `libgnt` subdirectory that was
extracted to produce this repository. Unfortunately, we were not consistent
with updating the version in that copy. It was deemed too much work for little
gain to go back through the history to correct this error when producing this
split repository. As such, the third column of the table below contains the
version *as recorded in `configure.ac`* at the time of a tag.

Beginning with version 2.14.0, the code has been moved to a separate repository
and so the tag and the version will remain in sync.

Tagged   | Released   | Recorded
-------- | ---------- | ----------
v2.14.1  | 2.14.1     | 2.14.1
v2.14.0  | 2.14.0     | 2.14.0
v2.13.0  | 2.8.10     | 2.8.0devel
v2.12.0  | 2.8.10     | 2.8.0devel
v2.11.0  | 2.8.10     | 2.8.0devel
v2.10.12 | 2.8.10     | 2.8.0devel
v2.10.11 | 2.8.10     | 2.8.0devel
v2.10.10 | 2.8.10     | 2.8.0devel
v2.10.9  | 2.8.10     | 2.8.0devel
v2.10.8  | 2.8.10     | 2.8.0devel
v2.10.7  | 2.8.9      | 2.8.0devel
v2.10.6  | 2.8.9      | 2.8.0devel
v2.10.5  | 2.8.9      | 2.8.0devel
v2.10.4  | 2.8.9devel | 2.8.0devel
v2.10.3  | 2.8.9      | 2.8.0devel
v2.10.2  | 2.8.9      | 2.8.0devel
v2.10.1  | 2.8.9      | 2.8.0devel
v2.10.0  | 2.8.9      | 2.8.0devel
v2.9.0   | 2.8.9      | 2.8.0devel
v2.8.0   | 2.8.8      | 2.8.0devel
v2.7.11  | 2.8.7      | 2.8.0devel
v2.7.10  | 2.8.6      | 2.8.0devel
v2.7.9   | 2.8.5      | 2.8.0devel
v2.7.8   | 2.8.5      | 2.8.0devel
v2.7.7   | 2.8.4      | 2.8.0devel
v2.7.6   | 2.8.3      | 2.8.0devel
v2.7.5   | 2.8.2      | 2.8.0devel
v2.7.4   | 2.8.1      | 2.8.0devel
v2.7.3   | 2.8.0      | 2.8.0devel
v2.7.2   | 2.7.2      | 2.7.0devel
v2.7.1   | 2.7.1      | 2.7.0devel
v2.7.0   | 2.7.0      | 2.7.0devel
v2.6.6   | 2.6.6      | 2.6.2devel
v2.6.5   | 2.6.5      | 2.6.2devel
v2.6.4   | 2.6.4      | 2.6.2devel
v2.6.3   | 2.6.3      | 2.6.2devel
v2.6.2   | 2.6.2      | 2.6.2devel
v2.6.1   | 2.6.1      | 2.5.0
v2.6.0   | 2.6.0      | 2.5.0
v2.5.9   | 2.5.9      | 2.5.6
v2.5.8   | 2.5.8      | 2.5.6
v2.5.7   | 2.5.7      | 2.5.6
v2.5.6   | 2.5.6      | 2.5.6
v2.5.5   | 2.5.5      | 2.5.0
v2.5.4   | 2.5.4      | 2.5.0
v2.5.3a  | 2.5.3      | 2.5.0
v2.5.3   | 2.5.3devel | 2.5.0
v2.5.2   | 2.5.2      | 2.5.0
v2.5.1   | 2.5.1      | 2.5.0
v2.5.0   | 2.5.0      | 2.5.0
v2.4.3   | 2.4.3      | 2.4.2
v2.4.2   | 2.4.2      | 2.4.2
v2.4.1   | 2.4.1      | 2.4.1
v2.4.0   | 2.4.0      | 2.4.0devel
v2.3.1   | 2.3.1      | 2.3.1
v2.3.0   | 2.3.0      | 2.3.0
v2.2.2   | 2.2.2      | 2.2.0devel
v2.2.1   | 2.2.1      | 2.2.0devel
v2.2.0   | 2.2.0      | 2.2.0devel
v2.1.1   | 2.1.0      | 2.1.0devel
v2.1.0   | 2.0.0      | 2.0.0devel
v2.0.2   | 1.0.2      | 1.0.0beta7
v2.0.1   | 1.0.1      | 1.0.0beta7
v2.0.0   | 1.0.0      | 1.0.0beta7


ChangeLog
GNT: The GLib Ncurses Toolkit
=============================

Please see notes about versioning in README.md.

version 2.14.3 (2021/09/25):
	* Added an option to disable python2 support. (LIBGNT-16) (RR 930) (Gary
	  Kramlich)
	* Add an option to disable building the docs. (LIBGNT-17) (RR 945) (Helmut
	  Grohne, Gary Kramlich)

version 2.14.2 (2021/07/23):
	* Fix a SEGFAULT in gnt_combo_box_get_dropdown. (RR #793) (ivanhoe)
	* Fix an invalid read/write when hiding widgets. (RR #810) (ivanhoe)
	* Look for python-2.7.pc as well as python2.pc. (RR #822) (Gary Kramlich)

version 2.14.1 (2021/02/04):
	* Bump the minimum meson version from 0.37.0 to 0.41.0 (Richard Laager)
	* Fix a buffer size in gntwm.c.  (Richard Laager)

version 2.14.0 (2020/06/10):
	* Split libgnt out of the Pidgin repository. (Elliott Sales de Andrade)
	* Switched build system to Meson. (Elliott Sales de Andrade)
	* Fixed a build issue on FreeBSD. (Steve Wills)

version 2.13.0 (2018/03/08):
	* Fixed build against curses 6.0 with opaque structs set. (#16764 dimstar)
	  (PR #268 Daniel Kamil Kozar)
	* Fixed a crash when resizing the window. (#16680 marcus) (PR #269 Daniel Kamil Kozar)

version 2.12.0 (2017/03/09):
	No changes.

version 2.11.0 (2016/06/21):
	No changes.

version 2.10.12 (2015/12/31):
	No changes.

version 2.10.11 (2014/11/23):
	No changes.

version 2.10.10 (2014/10/22):
	* Fix build against Python 3. (Ed Catmur) (#15969)

version 2.10.9 (2014/02/02):
	No changes.

version 2.10.8 (2014/01/28):
	No changes.

version 2.10.7 (2013/02/13):
	Alien hatchery:
	* No changes

version 2.10.6 (2012/07/06):
	No changes.

version 2.10.5 (2012/07/05):
	No changes.

version 2.10.4 (2012/05/06):
	No changes.

version 2.10.3 (2012/03/26):
	No changes.

version 2.10.2 (2012/03/14):
	No changes.

version 2.10.1 (2011/12/06):
	* Fix compilation on OpenBSD.

version 2.10.0 (2011/08/18):
	No changes.

version 2.9.0 (2011/06/23):
	No changes.

version 2.8.0 (2011/06/07):
	No changes.

version 2.7.11 (2011/03/10):
	No changes.

version 2.7.10 (2011/02/06):
	No changes.

version 2.7.9 (2010/12/26):
	No changes.

version 2.7.8 (2010/12/19):
	No changes.

version 2.7.7 (2010/11/23):
	No changes.

version 2.7.6 (2010/11/21):
	No changes.

version 2.7.5 (2010/10/31):
	No changes.

version 2.7.4 (2010/10/20):
	No changes.

version 2.7.3 (2010/08/10):
	* Rebindable 'suggest-next-page' and 'suggest-prev-page' actions for
	  textboxes (GntEntry) to scroll through list of suggestions.
	* Rebindable 'dropdown' action for comboboxes (GntComboBox) to show the
	  dropdown list of options.

version 2.7.2 (2010/07/21):
	No changes.

version 2.7.1 (2010/05/29):
	No changes.

version 2.7.0 (2010/05/12):
	* Changed GLib minimum version requirement to 2.12.0.
	* Fix autoconf detection of Python.  (Brad Smith)
	* New action 'history-search', with default binding ctrl+r, to search
	  the entered string in the input history.

version 2.6.6 (2010/02/18):
	* Fix CVE-2010-0420, a possible remote crash when handling chat room
	  buddy names.
	* Rebindable 'move-first' and 'move-last' actions for tree widgets. So
	  it is possible to jump to the first or last entry in the buddy list
	  (and other such lists) by pressing home or end key (defaults)
	  respectively.

version 2.6.5 (2010/01/08):
	* Build-time fixes for Solaris.  (Paul Townsend)

version 2.6.4 (2009/11/29):
	* Fix displaying umlauts etc. in non-utf8 locale.

version 2.6.3 (2009/10/16):
	No changes.

version 2.6.2 (2009/09/05):
	* Properly detect libpanel on OpenBSD.  (Brad Smith)
	* Remove IO watches in gnt_quit.  (Tomasz Mon)

version 2.6.1 (2009/08/18):
	No changes.

version 2.6.0 (2009/08/18):
	* The hardware cursor is updated correctly. This will be useful
	  especially for users of braille terminals, screen readers etc.
	* Fixed UTF-8 compatibility problems which could cause exits or other
	  unrequested behaviour.
	* Add a new progress bar widget.  (Saleem Abdulrasool)

version 2.5.9 (2009/08/18):
	No changes.

version 2.5.8 (2009/06/27):
	No changes.

version 2.5.7 (2009/06/20):
	No changes.

version 2.5.6 (2009/05/19):
	No changes.

version 2.5.5 (2009/03/01):
	* Allow rebinding keys to change the focused widget (details in the
	  man-page, look for GntBox::binding)

version 2.5.4 (2009/01/12):
	* Fix rebinding actions with the arrow-keys and tab.

version 2.5.3 (2008/12/20):
	* Allow binding meta+arrow keys for actions.
	* Added default meta+erase binding for delete previous word.

version 2.5.2 (2008/10/19):
	* A bug was fixed where some key-bindings wouldn't work with some TERMs
	  (e.g. xterm-color, screen-linux etc.)

version 2.5.1 (2008/08/30):
	No changes.

version 2.5.0 (2008/08/18):
	No changes.

version 2.4.2 (2008/05/17):
	* New default binding ctrl+x to open context menus.
	* Menu triggers and other bindings will no longer conflict.
	* Middle click pastes the internal clipboard (when mouse support is
	  enabled).

version 2.4.1 (2008/03/31):
	* Fix compiling with Glib older than 2.6

version 2.4.0 (2008/02/29):
	* The default keybinding for dump-screen is now M-D and uses a file
	  request dialog. M-d will properly delete-forward-word, and M-f has been
	  fixed to imitate readline's behavior.
	* New bindings alt+tab and alt+shift+tab to help navigating between the
	  higlighted windows (details on the man page).

version 2.3.0 (2007/11/24):
	* It's possible to bind key-strokes to specific menuitems in the windows.
	  Read the 'Menus' section in the man-page for details.
	* 'transpose-chars' operation for the entry boxes. The default key-binding
	  is ctrl+t.
	* 'yank' operation for the entry boxes. The default binding is ctrl+y.

version 2.2.2 (2007/10/23):
	* Various bug and memory leak fixes

version 2.2.1 (2007/09/29):
	No changes.

version 2.2.0 (2007/09/13):
	No changes.

version 2.1.1 (2007/08/20):
	No changes.

version 2.1.0 (2007/07/28):
	* There's support for workspaces now (details in the manpage)
	* There's a new custom window manager, Irssi

gntkeys.c
/*
 * GNT - The GLib Ncurses Toolkit
 *
 * GNT is the legal property of its developers, whose names are too numerous
 * to list here.  Please refer to the COPYRIGHT file distributed with this
 * source distribution.
 *
 * This library is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02111-1301  USA
 */

#include "gntinternal.h"
#undef GNT_LOG_DOMAIN
#define GNT_LOG_DOMAIN "Keys"

#include "gntkeys.h"

#include <glib.h>
#include <stdlib.h>
#include <string.h>

char *gnt_key_cup;
char *gnt_key_cdown;
char *gnt_key_cleft;
char *gnt_key_cright;

static const char *term;
static GHashTable *specials;

void gnt_init_keys()
{
	const char *controls[] = {"", "c-", "ctrl-", "ctr-", "ctl-", NULL};
	const char *alts[] = {"", "alt-", "a-", "m-", "meta-", NULL};
	int c, a, ch;
	char key[32];

	if (term == NULL) {
		term = getenv("TERM");
		if (!term)
			term = "";  /* Just in case */
	}

	if (strstr(term, "xterm") == term || strcmp(term, "rxvt") == 0) {
		gnt_key_cup    = "\033" "[1;5A";
		gnt_key_cdown  = "\033" "[1;5B";
		gnt_key_cright = "\033" "[1;5C";
		gnt_key_cleft  = "\033" "[1;5D";
	} else if (strstr(term, "screen") == term || strcmp(term, "rxvt-unicode") == 0) {
		gnt_key_cup    = "\033" "Oa";
		gnt_key_cdown  = "\033" "Ob";
		gnt_key_cright = "\033" "Oc";
		gnt_key_cleft  = "\033" "Od";
	}

	specials = g_hash_table_new(g_str_hash, g_str_equal);

#define INSERT_KEY(k, code) do { \
		g_hash_table_insert(specials, g_strdup(k), g_strdup(code)); \
		gnt_keys_add_combination(code); \
	} while (0)

	INSERT_KEY("home",     GNT_KEY_HOME);
	INSERT_KEY("end",      GNT_KEY_END);
	INSERT_KEY("pageup",   GNT_KEY_PGUP);
	INSERT_KEY("pagedown", GNT_KEY_PGDOWN);
	INSERT_KEY("insert",   GNT_KEY_INS);
	INSERT_KEY("delete",   GNT_KEY_DEL);
	INSERT_KEY("back_tab", GNT_KEY_BACK_TAB);

	INSERT_KEY("left",   GNT_KEY_LEFT);
	INSERT_KEY("right",  GNT_KEY_RIGHT);
	INSERT_KEY("up",     GNT_KEY_UP);
	INSERT_KEY("down",   GNT_KEY_DOWN);

	INSERT_KEY("tab",    "\t");
	INSERT_KEY("escape", "\033");
	INSERT_KEY("space", " ");
	INSERT_KEY("return", GNT_KEY_ENTER);
	INSERT_KEY("menu",   GNT_KEY_POPUP);

	INSERT_KEY("f1",   GNT_KEY_F1);
	INSERT_KEY("f2",   GNT_KEY_F2);
	INSERT_KEY("f3",   GNT_KEY_F3);
	INSERT_KEY("f4",   GNT_KEY_F4);
	INSERT_KEY("f5",   GNT_KEY_F5);
	INSERT_KEY("f6",   GNT_KEY_F6);
	INSERT_KEY("f7",   GNT_KEY_F7);
	INSERT_KEY("f8",   GNT_KEY_F8);
	INSERT_KEY("f9",   GNT_KEY_F9);
	INSERT_KEY("f10",  GNT_KEY_F10);
	INSERT_KEY("f11",  GNT_KEY_F11);
	INSERT_KEY("f12",  GNT_KEY_F12);

#define REM_LENGTH  (sizeof(key) - (cur - key))
#define INSERT_COMB(k, code) do { \
		snprintf(key, sizeof(key), "%s%s%s", controls[c], alts[a], k);  \
		INSERT_KEY(key, code);  \
	} while (0)
#define INSERT_COMB_CODE(k, c1, c2) do { \
		char __[32]; \
		snprintf(__, sizeof(__), "%s%s", c1, c2); \
		INSERT_COMB(k, __); \
	} while (0)

	/* Lower-case alphabets */
	for (a = 0, c = 0; controls[c]; c++, a = 0) {
		if (c) {
			INSERT_COMB("up",    gnt_key_cup);
			INSERT_COMB("down",  gnt_key_cdown);
			INSERT_COMB("left",  gnt_key_cleft);
			INSERT_COMB("right", gnt_key_cright);
		}

		for (a = 0; alts[a]; a++) {
			for (ch = 0; ch < 26; ch++) {
				char str[2] = {'a' + ch, 0}, code[4] = "\0\0\0\0";
				int ind = 0;
				if (a)
					code[ind++] = '\033';
				code[ind] = (c ? 1 : 'a') + ch;
				INSERT_COMB(str, code);
			}
			if (c == 0 && a) {
				INSERT_COMB("tab", "\033\t");
				INSERT_COMB_CODE("up", "\033", GNT_KEY_UP);
				INSERT_COMB_CODE("down", "\033", GNT_KEY_DOWN);
				INSERT_COMB_CODE("left", "\033", GNT_KEY_LEFT);
				INSERT_COMB_CODE("right", "\033", GNT_KEY_RIGHT);
			}
		}
	}
	c = 0;
	for (a = 0; alts[a]; a++) {
		/* Upper-case alphabets */
		for (ch = 0; ch < 26; ch++) {
			char str[2] = {'A' + ch, 0}, code[] = {'\033', 'A' + ch, 0};
			INSERT_COMB(str, code);
		}
		/* Digits */
		for (ch = 0; ch < 10; ch++) {
			char str[2] = {'0' + ch, 0}, code[] = {'\033', '0' + ch, 0};
			INSERT_COMB(str, code);
		}
	}
}

void gnt_keys_refine(char *text)
{
	while (*text == 27 && *(text + 1) == 27)
		text++;
	if (*text == 27 && *(text + 1) == '[' &&
			(*(text + 2) >= 'A' && *(text + 2) <= 'D')) {
		/* Apparently this is necessary for urxvt and screen and xterm */
		if (strstr(term, "screen") == term || strcmp(term, "rxvt-unicode") == 0 ||
				strstr(term, "xterm") == term ||
				strstr(term, "vt100") == term)
			*(text + 1) = 'O';
	} else if (g_utf8_get_char(text) == 195) {
		if (*(text + 2) == 0 && strstr(term, "xterm") == term) {
			*(text) = 27;
			*(text + 1) -= 64;  /* Say wha? */
		}
	}
}

const char *gnt_key_translate(const char *name)
{
	return name ? g_hash_table_lookup(specials, name) : NULL;
}

typedef struct {
	const char *name;
	const char *key;
} gntkey;

static void
get_key_name(gpointer key, gpointer value, gpointer data)
{
	gntkey *k = data;
	if (k->name)
		return;
	if (g_utf8_collate(value, k->key) == 0)
		k->name = key;
}

const char *gnt_key_lookup(const char *key)
{
	gntkey k = {NULL, key};
	g_hash_table_foreach(specials, get_key_name, &k);
	return k.name;
}

/*
 * The key-bindings will be saved in a tree. When a keystroke happens, GNT will
 * find the sequence that matches a binding and return the length.
 * A sequence should not be a prefix of another sequence. If it is, then only
 * the shortest one will be processed. If we want to change that, we will need
 * to allow getting the k-th prefix that matches the input, and pay attention
 * to the return value of gnt_wm_process_input in gntmain.c.
 */
#define SIZE 256

#define IS_END         (1 << 0)
struct _node
{
	struct _node *next[SIZE];
	int ref;
	int flags;
};

static struct _node root = {.ref = 1, .flags = 0};

static void add_path(struct _node *node, const char *path)
{
	struct _node *n = NULL;
	if (!path || !*path) {
		node->flags |= IS_END;
		return;
	}
	while (*path && node->next[(unsigned char)*path]) {
		node = node->next[(unsigned char)*path];
		node->ref++;
		path++;
	}
	if (!*path)
		return;
	n = g_new0(struct _node, 1);
	n->ref = 1;
	node->next[(unsigned char)*path++] = n;
	add_path(n, path);
}

void gnt_keys_add_combination(const char *path)
{
	add_path(&root, path);
}

static void del_path(struct _node *node, const char *path)
{
	struct _node *next = NULL;

	if (!*path)
		return;
	next = node->next[(unsigned char)*path];
	if (!next)
		return;
	del_path(next, path + 1);
	next->ref--;
	if (next->ref == 0) {
		node->next[(unsigned char)*path] = NULL;
		g_free(next);
	}
}

void gnt_keys_del_combination(const char *path)
{
	del_path(&root, path);
}

int gnt_keys_find_combination(const char *path)
{
	int depth = 0;
	struct _node *n = &root;

	root.flags &= ~IS_END;
	while (*path && n->next[(unsigned char)*path] && !(n->flags & IS_END)) {
		if (!g_ascii_isspace(*path) &&
				!g_ascii_iscntrl(*path) &&
				!g_ascii_isgraph(*path))
			return 0;
		n = n->next[(unsigned char)*path++];
		depth++;
	}

	if (!(n->flags & IS_END))
		depth = 0;
	return depth;
}

static void
print_path(struct _node *node, int depth)
{
	int i;
	for (i = 0; i < SIZE; i++) {
		if (node->next[i]) {
			g_printerr("%*c (%d:%d)\n", depth * 4, i, node->next[i]->ref,
						node->next[i]->flags);
			print_path(node->next[i], depth + 1);
		}
	}
}

/* this is purely for debugging purposes. */
void gnt_keys_print_combinations(void);
void gnt_keys_print_combinations()
{
	g_printerr("--------\n");
	print_path(&root, 1);
	g_printerr("--------\n");
}

